packages/workspace/src/core/affected-project-graph/locators/workspace-json-changes.ts

```diff
@@ -0,0 +1,49 @@
+import {
+  isWholeFileChange,
+  WholeFileChange,
+  workspaceFileName
+} from '../../file-utils';
+import { isJsonChange, JsonChange } from '../../../utils/json-diff';
+import { TouchedProjectLocator } from '../affected-project-graph-models';
+
+export const getTouchedProjectsInWorkspaceJson: TouchedProjectLocator<
+  WholeFileChange | JsonChange
+> = (touchedFiles, workspaceJson): string[] => {
+  const workspaceChange = touchedFiles.find(
+    change => change.file === workspaceFileName()
+  );
+  if (!workspaceChange) {
+    return [];
+  }
+
+  const changes = workspaceChange.getChanges();
+
+  if (
+    changes.some(change => {
+      if (isJsonChange(change)) {
+        return change.path[0] !== 'projects';
+      }
+      if (isWholeFileChange(change)) {
+        return true;
+      }
+      return false;
+    })
+  ) {
+    return Object.keys(workspaceJson.projects);
+  }
+
+  const touched = [];
+  changes.forEach(change => {
+    if (!isJsonChange(change) || change.path[0] !== 'projects') {
+      return;
+    }
+
+    if (workspaceJson.projects[change.path[1]]) {
+      touched.push(change.path[1]);
+    } else {
+      // The project was deleted so affect all projects
+      touched.push(...Object.keys(workspaceJson.projects));
```

Comment:
The `forEach` iteration can be skipped once we know all projects have been touched. I did the same thing in [here](https://github.com/nrwl/nx/blob/master/packages/workspace/src/core/affected-project-graph/locators/npm-packages.ts#L20-L25), but later realized it could be optimized.